File: frontend/src/services/SkillAnalyticsService.js
import apiClient from './apiClient';

const ANALYTICS_URL = '/skill-analytics';

export const SkillAnalyticsService = {
  // Recalculate popularity score for a skill
  recalculateSkillPopularity: async (skillId) => {
    try {
      await apiClient.post(`${ANALYTICS_URL}/recalculate/${skillId}`);
    } catch (error) {
      console.error(`Error recalculating popularity for skill ${skillId}:`, error);
      throw error;
    }
  },

  // Identify trending skills
  identifyTrendingSkills: async () => {
    try {
      const response = await apiClient.post(`${ANALYTICS_URL}/identify-trending`);
      return response.data;
    } catch (error) {
      console.error('Error identifying trending skills:', error);
      throw error;
    }
  },

  // Get trending skills with pagination
  getTrendingSkills: async (page = 0, size = 10) => {
    try {
      const response = await apiClient.get(`${ANALYTICS_URL}/trending`, {
        params: { page, size }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching trending skills:', error);
      throw error;
    }
  },

  // Track skill usage
  trackSkillUsage: async (skillId) => {
    try {
      await apiClient.post(`${ANALYTICS_URL}/track-usage/${skillId}`);
    } catch (error) {
      console.error(`Error tracking usage for skill ${skillId}:`, error);
      throw error;
    }
  },

  // Get most frequent tags
  getMostFrequentTags: async (limit = 10) => {
    try {
      const response = await apiClient.get(`${ANALYTICS_URL}/tags/most-frequent`, {
        params: { limit }
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching most frequent tags:', error);
      throw error;
    }
  },

  // Get popularity by category
  getPopularityByCategory: async () => {
    try {
      const response = await apiClient.get(`${ANALYTICS_URL}/popularity-by-category`);
      return response.data;
    } catch (error) {
      console.error('Error fetching popularity by category:', error);
      throw error;
    }
  },

  // Get skill distribution by difficulty
  getSkillDistributionByDifficulty: async () => {
    try {
      const response = await apiClient.get(`${ANALYTICS_URL}/distribution-by-difficulty`);
      return response.data;
    } catch (error) {
      console.error('Error fetching skill distribution by difficulty:', error);
      throw error;
    }
  }
};

export default SkillAnalyticsService;

File: backend/src/main/java/com/ds39/mastermind/service/SkillAnalyticsService.java
package com.ds39.mastermind.service;

import com.ds39.mastermind.dto.skill.SkillDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Map;

public interface SkillAnalyticsService {
    
    /**
     * Recalculates the popularity score for a skill
     * @param skillId The skill ID to update
     */
    void recalculateSkillPopularity(Long skillId);
    
    /**
     * Identifies trending skills based on recent usage and popularity
     * @return List of trending skill IDs
     */
    List<Long> identifyTrendingSkills();
    
    /**
     * Gets trending skills with pagination
     * @param pageable Pagination information
     * @return Page of trending skills
     */
    Page<SkillDTO> getTrendingSkills(Pageable pageable);
    
    /**
     * Tracks the usage of a skill (increments count)
     * @param skillId The skill ID to track
     */
    void trackSkillUsage(Long skillId);
    
    /**
     * Gets the most frequently used skill tags
     * @param limit Maximum number of tags to return
     * @return Map of tag names to usage counts
     */
    Map<String, Long> getMostFrequentTags(int limit);
    
    /**
     * Gets skill popularity statistics by category
     * @return Map of category names to average popularity scores
     */
    Map<String, Double> getPopularityByCategory();
    
    /**
     * Gets skill distribution by difficulty level
     * @return Map of difficulty levels to counts
     */
    Map<String, Long> getSkillDistributionByDifficulty();
}